implementations/rust/ockam/ockam_transport_tcp/src/workers/listener.rs

```diff
@@ -54,12 +54,23 @@ impl Processor for TcpListenProcessor {
         debug!("TCP connection accepted");
 
         let handle_clone = self.router_handle.async_try_clone().await?;
-        // And spawn a connection worker for it
-        let pair = TcpSendWorker::start_pair(ctx, handle_clone, Some(stream), peer, vec![]).await?;
+        // And create a connection worker for it
+        let (worker, pair) =
+            TcpSendWorker::new_pair(ctx, handle_clone, Some(stream), peer, Vec::new()).await?;
 
         // Register the connection with the local TcpRouter
         self.router_handle.register(&pair).await?;
```

Comment:
Doesn't that change bring a new race condition? We registered the fact that our worker pair accepts messages for certain TCP addresses, but no mailbox has yet been created for them, so if something comes for that TCP address in that moment - it'll get lost. Although, I agree that this race condition may be less likely to happen and less harmful as the one fixed, but would be great to fix both

Reply:
Merging this PR and creating issue for a new race condition is an option

Reply:
Who could know about this socket address and send a message to it? The socket has just been accepted and the socket address contains an ephemeral port number.

Reply:
In the circumstances that @SanjoDeundiak described, a worker could only send a message to the socket address by chance, i.e. by trying random socket addresses as only the TCP router knows about this address at this point. Unless I am missing something this does not seem to be relevant. NB that the address we are looking at is the remote client address, not the remote listener address.

Reply:
@twittner yes, you're right. This newly accepted TCP address is in fact unknown and highly unlikely that someone will send a message to it, but it's still possible. 
I think the right approach is to avoid sharing the address, until there is the mailbox for it. Once you break that rule, it's tricky to verify that you don't have race conditions.
My suggestion is let's merge this PR and create an issue for the race condition, since it may take quite an effort to fix

Reply:
> This might be a race condition, but it _should_ be fine. From how I understand the issue, the series of events is as follows:
> 
> 1. Worker `A` sends a message to a TCP address
> 2. TCP router is set to "allow_auto_connect" and creates a new pair of workers
> 3. TCP router registers these workers as handling the respective TCP address
> 4. Messages routed to this address won't be delivered because the workers haven't been started yet
> 5. The connection workers are started.
> 
> But, because this logic happens in the message accept loop of the router, it won't get to any new message that it has to route, before the connection workers are (hopefully) initialised.
> 
> Alternatively we should be using the `context.wait_for` function mechanism instead to check that the connection workers are actually ready to receive and handle connections.

I think intended order of things is following:
1. Listener accepts the connection 
2. TCP worker pair is started
3. Worker pair is registered for the given TCP address
4. TCP receiver receives TCP traffic and forwards message to the node
5. Some worker replies to that message using return_route

While race condition happens if 3. is postponed due to async execution to be the last instruction, and in this case 5. fails (fails even more dramatically if auto_connect is enabled)

Reply:
> This newly accepted TCP address is in fact unknown and highly unlikely that someone will send a message to it, but it's still possible. I think the right approach is to avoid sharing the address, until there is a mailbox for it.

My point is that the address is in fact not shared but only part of the router's mapping.   This is not a race condition but more an issue of access control. Everyone can attempt to send something to any TCP address and potentially disrupt the protocol. Dropping such adversarial messages could even be considered a feature.

Reply:
Yes, but safer way is to keep address secret, until there is a mailbox ready receiving messages